Translate "How does the development mode work" into Spanish (#496)
gaearon/overreacted.io: src/pages/how-does-the-development-mode-work/index.es.md

---
title: ¿Cómo funciona el modo de desarrollo?
date: '2019-08-04'

> spoiler: Dead code elimination by convention.

Author: Ups! Corregido

La ralentización es aceptable en desarrollo. De hecho, ejecutar el código más lentamente en desarrollo *puede hasta ser beneficioso* porque compensa parcialmente la discrepancia entre máquinas rápidas de desarrollo y los dispositivos promedio de los consumidores.

En producción no queremos pagar ese costo. Por lo tanto, omitimos estas verificaciones. ¿Cómo funciona eso? Veamos.

---

La forma exacta para ejecutar código en desarrollo depende de tu *build pipeline* de JavaScript (y de si tienes una). En Facebook es algo así:

```js
if (__DEV__) {
  hacerAlgoEnDesarrollo();
} else {
  hacerAlgoEnProd();
}
```

Aquí, `__DEV__` no es una variable real. Es una constante que se sustituye cuando los módulos son compilados para el navegador. El resultado se ve así:

```js
// En desarrollo:
if (true) {
  hacerAlgoEnDesarrollo(); // 👈
} else {
  hacerAlgoEnProd();
}

// En production:
if (false) {
  hacerAlgoEnDesarrollo();
} else {
  hacerAlgoEnProd(); // 👈
}
```

> En producción, también ejecutarías un *minifier* (por ejemplo, [terser](https://github.com/terser-js/terser)) en el código. La mayoría de los *minifiers* de JavaScript hacen una forma limitada de [eliminación de código muerto](https://es.wikipedia.org/wiki/Eliminaci%C3%B3n_de_c%C3%B3digo_muerto), como quitar las ramas `if (false)`. Entonces en producción sólo verías:

Author: Perfecto, gracias. Lo mantengo en cursiva de todas formas, ¿te parece ok?

Hace unos años, solía ser común olvidarse configurar el entorno. Con frecuencia veías proyectos en modo de desarrollo desplegados en producción.

Eso es malo porque hace que el sitio web cargue y funcione más lentamente.

En los últimos dos años, la situación mejoró significativamente. Por ejemplo, webpack añadió una simple opción `mode` en lugar de tener que configurar manualmente el reemplazo de `process.env.NODE_ENV`. Ahora React DevTools también muestra un ícono rojo en sitios en modo de desarrollo, haciendo más fácil identificar e incluso [reportar](https://mobile.twitter.com/BestBuySupport/status/1027195363713736704).

![Advertencia de modo de desarrollo en React DevTools](devmode.png)

Instaladores dogmáticos como Create React App, Next/Nuxt, Vue CLI, Gatsby y otros hacen más difícil equivocarse con esto separando los *builds* de desarrollo y producción en dos comandos separados (por ejemplo, `npm start` y `npm run build`). Generalmente, solo se puede desplegar el build de producción por lo que el desarrollador ya no puede cometer este error.

> Siempre existe un argumento que quizás el modo *producción* necesita ser el valor por defecto y el modo de desarrollo opcional. Personalmente no me convence este argumento. Las personas que se benefician más de las advertencias en el modo de desarrollo generalmente son nuevos en la librería. *No sabrían cómo encenderlo,* y evitarían encontrar muchos bugs que con las advertencias se hacen evidentes.

Reviewer: Me resulta un poco extraño «encender» el modo de desarrollo. ¿Quizá mejor, «activarlo»?

Author: Aplicado!

> Si, los problemas de performance son malos, pero también es enviar experiencias rotas y llenas de bugs a los usuarios finales. Por ejemplo, el [warning de *key* de React](https://es.reactjs.org/docs/lists-and-keys.html#keys) ayuda a prevenir bugs como enviar un mensaje a la persona equivocada o comprar el producto erróneo. Desarrollar con esta advertencia desactivada es un riesgo significativo para tí *y* tus usuarios. Si está desactivada por defecto, entonces para el momento que encuentres la opción para encenderlo, tendrás demasiadas advertencias para limpiar. Así que muchas personas lo volverían a apagar. Este es el motivo por el que necesita estar encendido desde el comienzo en lugar de activado luego.

Reviewer: performance -> rendimiento
tí -> ti
```suggestion
Si, los problemas de rendimiento son malos, pero también es enviar experiencias rotas y llenas de bugs a los usuarios finales. Por ejemplo, el [warning de *key* de React](https://es.reactjs.org/docs/lists-and-keys.html#keys) ayuda a prevenir bugs como enviar un mensaje a la persona equivocada o comprar el producto erróneo. Desarrollar con esta advertencia desactivada es un riesgo significativo para ti *y* tus usuarios. Si está desactivada por defecto, entonces para el momento que encuentres la opción para encenderlo, tendrás demasiadas advertencias para limpiar. Así que muchas personas lo volverían a apagar. Este es el motivo por el que necesita estar encendido desde el comienzo en lugar de activado luego.
```

Author: Gracias! Fixed

Finalmente, incluso si las advertencias de desarrollo son opcionales y los desarrolladores *supieran* que deben encenderlas en una etapa temprana del desarrollo, simplemente volveríamos al mismo problema. ¡Alguien accidentalmente las dejaría encendidas en producción!

Y volvimos al primer casillero

> Personalmente creo en **herramientas que muestran y usan el modo correcto dependiendo si estás depurando o desplegando**. Casi cualquier entorno (sea mobile, escritorio o servidor) excepto el navegador ha tenido una forma de cargar y diferenciar builds de desarrollo y producción por décadas.

Reviewer: ```suggestion
Personalmente creo en **herramientas que muestran y usan el modo correcto dependiendo si estás depurando o desplegando**. Casi cualquier entorno (sea móvil, escritorio o servidor) excepto el navegador ha tenido una forma de cargar y diferenciar builds de desarrollo y producción por décadas.
```

La transformada original “envify” fue [lanzada en 2013](https://github.com/hughsk/envify/commit/ae8aa26b759cd2115eccbed96f70e7bbdceded97). React fue hecho open source por ese entonces y npm con browserify parecía la mejor solución para compilar código CommonJS de front-end.

React comenzó a proveer builds npm (adicionalmente a builds para etiquetas `<script>`) desde el comienzo. Cuando React se hizo popular, también lo hicieron las buenas prácticas para escribir JavaScript modular con CommonJS y entregar código de front-end vía npm.

React necesitaba quitar código que era sólo para desarrollo en el modo de producción. Browserify ya ofrecía una solución a este problema, así que React adoptó la convención de usar `process.env.NODE_ENV` para sus builds npm. Con el tiempo, muchas otras herramientas y librerías, incluyendo webpack y Vue, hicieron lo mismo.

> Por el 2019, browserify ha perdido bastante popularidad. Sin embargo, reemplazar `process.env.NODE_ENV` con `development` o `producción` en el paso de compilación es una convención aún popular.

Reviewer: ```suggestion
Por el 2019, browserify ha perdido bastante popularidad. Sin embargo, reemplazar `process.env.NODE_ENV` con `development` o `production` en el paso de compilación es una convención aún popular.
```

Desde React 16 cambiamos este enfoque. En su lugar, ahora [compilamos un bundle](https://reactjs.org/blog/2017/12/15/improving-the-repository-infrastructure.html#compiling-flat-bundles) para cada entorno (incluyendo las etiquetas`<script>`, npm y el código interno de Facebook). Así que incluso código en CommonJS para npm es compilado para separar bundles de desarrollo y de producción antes de tiempo.

Esto significa que mientras el código fuente de React dice `if (__DEV__)`, en realidad producimos dos bundles para cada paquete. Uno ya precompilado con `__DEV__ = true` y otro precompilado con `__DEV__ = false`. El punto de entrada para cada paquete en npm “decide” cuál exportar.

[Por ejemplo:](https://unpkg.com/browse/react@16.8.6/index.js)

```js
if (process.env.NODE_ENV === 'production') {
  module.exports = require('./cjs/react.production.min.js');
} else {
  module.exports = require('./cjs/react.development.js');
}
```

Y ese es el único lugar donde tu *bundler* interpolará `'development'` o `'production'` como un string, y donde tu minifier se deshacerá de ese `require` sólo para desarrollo.

> Tanto `react.production.min.js` como `react.development.js` ya no tienen verificaciones con `process.env.NODE_ENV`. Esto es muy bueno porque *cuando de hecho corremos en Node.js*, acceder a `process.env` es un [poco lento](https://reactjs.org/blog/2017/09/26/react-v16.0.html#better-server-side-rendering). compilar los bundles en ambas formas antes de tiempo también nos deja optimizar el tamaño de archivo de forma [mucho más consistente](https://reactjs.org/blog/2017/09/26/react-v16.0.html#reduced-file-size), independientemente de qué *bundler* o *minifier* uses.

Reviewer: compilar -> Compilar (en inicio de oración)
```suggestion
Tanto `react.production.min.js` como `react.development.js` ya no tienen verificaciones con `process.env.NODE_ENV`. Esto es muy bueno porque *cuando de hecho corremos en Node.js*, acceder a `process.env` es un [poco lento](https://reactjs.org/blog/2017/09/26/react-v16.0.html#better-server-side-rendering). Compilar los bundles en ambas formas antes de tiempo también nos deja optimizar el tamaño de archivo de forma [mucho más consistente](https://reactjs.org/blog/2017/09/26/react-v16.0.html#reduced-file-size), independientemente de qué *bundler* o *minifier* uses.
```
